src/Disks/DiskOverlay.cpp

```diff
@@ -0,0 +1,774 @@
+#include "DiskOverlay.h"
+
+#include <Disks/DiskFactory.h>
+#include <Disks/DirectoryIterator.h>
+#include <Disks/IDisk.h>
+#include <Disks/ObjectStorages/IObjectStorage_fwd.h>
+#include <Disks/ObjectStorages/MetadataStorageFactory.h>
+
+namespace DB
+{
+
+String dataPath(String path)
+{
+    if (!path.empty() && path.back() == '/')
+    {
+        path.pop_back();
+    }
+    path += ".data";
+    for (char & i : path)
+    {
+        if (i == '/')
+        {
+            i = '_';
+        }
+    }
+    return path;
+}
+
+String mergePath(String parent, const String& child)
+{
+    if (!parent.empty() && parent.back() != '/')
+    {
+        parent += '/';
+    }
+    parent += child;
+    return parent;
+}
+
+namespace ErrorCodes
+{
+    extern const int BAD_ARGUMENTS;
+    extern const int DIRECTORY_DOESNT_EXIST;
+    extern const int FILE_DOESNT_EXIST;
+    extern const int NOT_IMPLEMENTED;
+}
+
+DiskOverlay::DiskOverlay(const String & name_, DiskPtr disk_base_, DiskPtr disk_diff_, MetadataStoragePtr metadata_, MetadataStoragePtr tracked_metadata_) : IDisk(name_),
+disk_base(disk_base_), disk_diff(disk_diff_), forward_metadata(metadata_), tracked_metadata(tracked_metadata_)
+{
+}
+
+DiskOverlay::DiskOverlay(const String & name_, const Poco::Util::AbstractConfiguration & config_, const String & config_prefix_, const DisksMap & map_) : IDisk(name_)
+{
+    String disk_base_name = config_.getString(config_prefix_ + ".disk_base");
+    String disk_diff_name = config_.getString(config_prefix_ + ".disk_diff");
+    String metadata_name = config_.getString(config_prefix_ + ".metadata");
+    String tracked_metadata_name = config_.getString(config_prefix_ + ".tracked_metadata");
+
+    disk_base = map_.at(disk_base_name);
+    disk_diff = map_.at(disk_diff_name);
+
+    if (disk_diff -> isReadOnly())
+    {
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Diff disk has to be writable");
+    }
+
+    forward_metadata = MetadataStorageFactory::instance().create(metadata_name, config_, config_prefix_, nullptr, "");
```

Comment:
Do we support all types of disks? Let's add tests for different types / metadata types. A stateless test with dynamically created disk or an integration test will do. Probably an integration test is better, because would be also good to test server restarts for these disks.